satpy/resample/base.py

```diff
@@ -0,0 +1,217 @@
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+"""Base resampling functionality."""
+import hashlib
+import json
+import warnings
+from logging import getLogger
+from weakref import WeakValueDictionary
+
+import numpy as np
+from pyresample.resampler import BaseResampler as PRBaseResampler
+
+from satpy.utils import get_legacy_chunk_size
+
+LOG = getLogger(__name__)
+
+CHUNK_SIZE = get_legacy_chunk_size()
+
+resamplers_cache: "WeakValueDictionary[tuple, object]" = WeakValueDictionary()
+
+
+def _hash_dict(the_dict, the_hash=None):
+    """Calculate a hash for a dictionary."""
+    if the_hash is None:
+        the_hash = hashlib.sha1()  # nosec
+    the_hash.update(json.dumps(the_dict, sort_keys=True).encode("utf-8"))
+    return the_hash
+
+
+def _update_resampled_coords(old_data, new_data, new_area):
+    """Add coordinate information to newly resampled DataArray.
+
+    Args:
+        old_data (xarray.DataArray): Old data before resampling.
+        new_data (xarray.DataArray): New data after resampling.
+        new_area (pyresample.geometry.BaseDefinition): Area definition
+            for the newly resampled data.
+
+    """
+    from satpy.coords_utils import add_crs_xy_coords
+
+    # copy over other non-x/y coordinates
+    # this *MUST* happen before we set 'crs' below otherwise any 'crs'
+    # coordinate in the coordinate variables we are copying will overwrite the
+    # 'crs' coordinate we just assigned to the data
+    ignore_coords = ("y", "x", "crs")
+    new_coords = {}
+    for cname, cval in old_data.coords.items():
+        # we don't want coordinates that depended on the old x/y dimensions
+        has_ignored_dims = any(dim in cval.dims for dim in ignore_coords)
+        if cname in ignore_coords or has_ignored_dims:
+            continue
+        new_coords[cname] = cval
+    new_data = new_data.assign_coords(**new_coords)
+
+    # add crs, x, and y coordinates
+    new_data = add_crs_xy_coords(new_data, new_area)
+    return new_data
+
+
+# TODO: move this to pyresample.resampler
+#RESAMPLERS = {"kd_tree": KDTreeResampler,
+#              "nearest": KDTreeResampler,
+#              "bilinear": BilinearResampler,
+#              "native": NativeResampler,
+#              "gradient_search": create_gradient_search_resampler,
+#              "bucket_avg": BucketAvg,
+#              "bucket_sum": BucketSum,
+#              "bucket_count": BucketCount,
+#              "bucket_fraction": BucketFraction,
+#              "ewa": DaskEWAResampler,
+#              "ewa_legacy": LegacyDaskEWAResampler,
+#              }
+
+def get_all_resampler_classes():
+    """Get all available resampler classes."""
+    resamplers = {}
+    try:
+        from .native import get_native_resampler_classes
+        resamplers.update(get_native_resampler_classes())
+    except ImportError:
+        pass
```

Comment:
I don't think it needs to be done in this PR but this seems like another case where it would be easy to have a dictionary or list of strings to import and use `importlib.import_module` and `getattr` to get access to them. You could also use `with suppress(ImportError):` to avoid the "uglier" `except ImportError`.

Reply:
Refactored in https://github.com/pytroll/satpy/pull/3124/commits/9f9a3854c8d2c5d92a04d0e597c194f450245ecb . As a side-effect, all the resampler classes could be moved to Pyresample.